### DISCUSSION TITLE
[Idea] Unit tests

### DISCUSSION THREAD

**Discussion by Patrickclicweb** [PrestaShop/PrestaShop#31021] [Ideas]
### The problem

Unit tests often cause the abandonment of a PR because the contributor does not know / does not have the competence to do the required unit tests 

### The solution

For me I think that the management of PR and others being validated by the Prestashop team, it would be logical that Prestashop takes care of this kind of thing and why not integrate third party DEv as for the maintainers
So having a reactive support that responds and moves in the right direction would be a real plus

### Do you plan to work on this subject?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [ ] I'm willing to submit a Pull Request that implements this feature.
- [ ] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by kpodemski** [comment]
Hi Patrick

I think we need to be pragmatic with this one. Personally, I wouldn't block PR because of lack of unit tests if I couldn't provide a valid instruction how to handle them. I know it's a complex topic for many contributors and I wouldn't want some of our active contributors to go away because of that.

The truth is that many contributors learned how to write tests thanks to contributing, @Hlavtox @okom3pom and many others. How? They had a look at other tests that are already there.

If I can point to the right direction and a contributor is still telling me "I won't do it" then, well, we have to think twice because pull request is not a gift :)

**Reply by Patrickclicweb** [reply]
Hi Kristian !
Unit tests are a real concern for many contributors 
Already when we do a PR we give our time to the OpenSource project and that's good.
But to see like some examples 5 commits and all closed because no unit tests while the solution is conforma but the 5 contributors have solved the problem.
We have a real dialogue of deaf on Friend Of Presta on this 

And I come back to my problem, but it is more general
I give a PR 
I'm stuck because I don't know how to do unit tests, this one remains abandoned or is closed :( 
I know I can ask for help because we are a great group on FOP, but do I have the time, and the skills? 
This is a question I will ask at the next FOPDay ;) because it requires a real commitment from Prestashop SA because even if Prestashop SA is different from Prestashop Open Source it is still this company that has the hand on this kind of problem

**Reply by matks** [reply]
@Patrickclicweb If you dont have the skills, you can learn them. It's not hard.

There was a long time ago you did not know PHP. Did you ask someone else to write PHP instead of you? No, you learnt it. One day you did not know it. Next day you learnt. Some weeks later: hop! you know it.

When I started writing code in 2012 I was a Java developer. I did not know PHP. Well, I learnt it. And here I am. As far as I know coding is for everyone: everyone can learn any software language, and unit testing is a [standard development practice](https://standards.ieee.org/ieee/1008/1462/) in the coding world. If thousand other developers learnt to do it, why cannot you do it?

Don't think of yourself "less capable" of them 😊 . I insist: you _can_. It's not hard. You know git and HTML, right? It's not rocket science 😊 . Actually if you have learnt CSS I think unit tests are _easier_ 😋 . Unit testing is easy, trust me. It can look scary but it is not. You can learn it [by reading just this page](https://www.freecodecamp.org/news/test-php-code-with-phpunit/): I think we can agree HTML is more complex, isn't it?

There are blog articles, there are [YouTube videos](https://www.youtube.com/watch?v=9-X_b_fxmRM), there are hundred of material to help you. Simply google "learn write unit tests phpunit".

And if you think your unit test is not good? It's OK: you can submit a Pull Request with the unit test and reviewers will tell you if something is missing. Code review will guide you until the test is cool.

**Reply by andromaque** [reply]
@matks I disagree with your approach. PrestaShop (the open source project) should be glad that people are contributing and should not discourage them in these contributions, and PrestaShop SA has a huge interest of having contributions on the PrestaShop Open Source project as it benefits from these improvements to sell addons, hosting, insurances etc... It's an evidence that the knowledge of unit tests is available to only a minority. Of course people can learn. But is it really always needed?

Furthermore, it would make a lot of sense to have a dedicated team to the writing of unit tests, because
1. it's important to build a proper validation plan
2. it would ensure a consistency and integrity amongst all the tests
3. it would allow more contributions to the projects.

Finally, specifying a change, coding it and writing a QC are 3 different things that should be handled (possibly) by different people. It's great if someone can do both of them, but it should not be mandatory.

**Comment by prettytree** [comment]
i am send hugs to the prestashop team that receive that

**Comment by matks** [comment]
Hello there ✋ this discussion did not move since January so I think the people who opened it and read it had nothing to add.

I just wanted to provide at least a final word (for my side):

I can't speak for the whole maintainer group so I only voice my opinion: I am sorry I reject your idea. I have been trained my whole developer life (10 years) with the idea that "you write the code and the tests together, one cannot live without the other" and I think it really is important the same person who modifies some code adds the needed tests for it. It should be mandatory that the person who writes some code and the person who writes the unit test that validates it are the same person in my opinion.

However I am just one maintainer of PrestaShop: if another maintainer disagrees with me he can step up and start discussing with you on how to do it.

I will not write your unit tests for you. _However_
- I will be happy providing you with guides and how-to to write unit tests
- I will be happy giving you hints and advices on how to write unit tests
- I will be happy to review your Pull Request that contain unit tests and tell you whether they are good or they miss something (and if they miss something I will tell you what)

I think you know this quote:
> “If you give a man a fish, you feed him for a day. If you teach a man to fish, you feed him for a lifetime.”

**Comment by leemyongpakvn** [comment]
I'm not against **matks** in this case. But from **Patrickclicweb** and **andromaque** side, it seems if they want to contribute a fish, they need to contribute a pan and a cooker too. :joy:

**Comment by andromaque** [comment]
I think that @leemyongpakvn summarized it very well 😄

Unfortunately there is no participation in this topic, so there is not much of a debate here, rather two different opinions expressed. You disagree with my arguments, I disagree with yours, it's ok but it won't make things better at the end.

I still strongly believe that unit tests / QC plan should be performed by a different people / unit, it's not only a different skill, it's also a good way to ensure the quality of the programming. If you programmed badly, it's safe to assume that you will also test badly, and with this system there are a lot of chances that it will happen.

**Reply by matks** [reply]
I agree the discussion is over, you have your opinion, I have mine 👍 .

However here is some more food for thoughts: can you tell me about a company where the people who write the code and the people who write the unit tests are two different teams?

Because I can probably mention a dozen companies where the people who write the code and the people who write the unit tests are the same team (and to be more accurate, companies where a developer who writes feature A also writes the unit tests for feature A).

**Comment by Patrickclicweb** [comment]
I agree with @andromaque 
We know how to solve some problems, we give solutions, we just don't have the time/knowledge or resources to do better, that's our way of contributing 
If you think that this discussion is useless then you will send a bad signal to the community

@matks 
I know several companies that do this like my brother-in-law's in the medical field (there are many companies that do it) because it is good for the bugg tracker and the vision of another team ;) 
I have less even a customer who develops custom software and the testing and bug tracking teams are not the same 

I also validated in another domain that to be controlled by others was beneficial to the company
I managed 5 sites including 3 Prestashop for a company with 80 physical stores
We had a lot of errors from the order pickers 
We set up a system which made that the team which prepared the order it was another team which made the forwarding, we reduced in qulques weeks of 75% the errors 

So just because you don't see companies doing it doesn't mean it's true. 
You have to know how to open up to the world, so that the world opens up to you  
close this discussion, it will save you work here is what I think of your vision 
best regards

**Reply by kpodemski** [reply]
Patrick, you don't want to do to them because you can't, ok. Would you like for us to guide you? Or do you not care? "Here's my big change, completely untested. Write tests if you want, guys". Do you think that's ok? As I have written several times. We have amazing people in the community who asked questions, discussed with other project members, and then participated in the project with fully tested PRs! @okom3pom @tom-combet @Hlavtox @ks129 and many, many others!

Please keep in mind that we are not fully strict on the tests, and you can participate in the project without writing them very often. Maybe it's not the best for some, but... ;-)

**Comment by Hlavtox** [comment]
@Patrickclicweb @andromaque Guys, I would not see it that critically. For many contributions, especially to the legacy parts, there are no tests needed.

If migrating something to symfony and it was testable, I always went and checked the current and found something simmilar, and somehow glued it together. And when you do a third PR, you already know it. It's not that big of a deal.

And if you are having trouble with something, there will be always be super skilled somebody who will lend you a hand, give you advice and tell you how to do it. Trust me. :-)
